File: c++/leetcode/538. Convert BST to Greater Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
  void solve(TreeNode *, int &);
  TreeNode *convertBST(TreeNode *root)
  {
    int sum = 0;
    solve(root, sum);
    return root;
  }
};

void Solution::solve(TreeNode *root, int &sum)
{
  if (!root)
    return;

  solve(root->right, sum);

  root->val += sum;
  sum = root->val;

  solve(root->left, sum);
}

File: javascript/leetcode/538. Convert BST to Greater Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var convertBST = function (root) {
  let sum = 0;

  const solve = (root) => {
    if (!root) return;

    solve(root.right);

    root.val += sum;
    sum = root.val;

    solve(root.left);
  };

  solve(root);
  return root;
};
